File: bench/mat/mat.py
class SimpleMatrix(object):

    def __init__(self, N, data):
        self._data = data
        self._N = N

    def __add__(self, other):
        N = self._N
        data = list(range(N*N))

        for i in range(N*N):
            data[i] = self._data[i] + other._data[i]

        return SimpleMatrix(N, data)

    def __mul__(self, other):
        N = self._N
        data = list(range(N*N))

        for i in range(N):
            for j in range(N):
                sum = 0
                for k in range(N):
                    sum += self._data[i*N+k] * other._data[k*N+j]
                data[i*N+j] = sum

        return SimpleMatrix(N, data)

    def diag_sum(self):
        N = self._N
        sum = 0

        for i in range(N):
            sum += self._data[i*(N+1)]

        return sum

sum = 0


N = 2


for i in range(5000):
    a = SimpleMatrix(N, [i for i in range(N * N)])
    b = SimpleMatrix(N, [i for i in range(N * N)])

    c = a + b
    d = c * a * b

    sum = d.diag_sum()

print(sum)

File: bench/mat/mat.rb
class SimpleMatrix
  attr_accessor :data

  def initialize(n, data)
    @data = data
    @N = n
  end

  def +(other)
    n = @N
    data = Array.new(n * n)

    (n*n).times do |i|
      data[i] = @data[i] + other.data[i]
    end

    return SimpleMatrix.new(n, data)
  end

  def *(other)
    n = @N
    data = Array.new(n * n)

    n.times do |i|
      n.times do |j|
        sum = 0

        n.times do |k|
          sum += @data[i*n+k] * other.data[k*n+j]
        end

        data[i*n+j] = sum
      end
    end
    
    return SimpleMatrix.new(n, data)
  end

  def diag_sum
    n = @N
    sum = 0

    n.times do |i|
      sum += @data[i*(n + 1)]
    end

    return sum
  end
end

sum = 0

n = 2

5000.times do |i|
  a = SimpleMatrix.new(n, Array.new(n * n).map { i | i })
  b = SimpleMatrix.new(n, Array.new(n * n).map { i | i })

  c = a + b
  d = c * a * b

  sum = d.diag_sum
end

print(sum)
